### DISCUSSION TITLE
error 429, but requests executes the request correctly

### DISCUSSION THREAD

**Discussion by AndrejedaI** [aio-libs/aiohttp#9298] [Q&A] [answered]
I'm trying to get a response from a server with a status of 200. The request library handles this, but aiohttp doesn't.
Perhaps the first 5-7 requests will give an answer of 200, but then it will be 429. It’s strange that it’s not clear what this could be connected with.

```python
import asyncio
import aiohttp
import requests


headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:130.0) Gecko/20100101 Firefox/130.0',
    'Accept': '*/*',
    'Accept-Language': 'ru-RU,ru;q=0.8,en-US;q=0.5,en;q=0.3',
    'Content-Type': 'text/plain;charset=UTF-8',
    'Origin': 'https://app.uniswap.org',
}

data = '{"tokenInChainId":1,"tokenIn":"ETH","tokenOutChainId":1,"tokenOut":"0xe41d2489571d322189246dafa5ebde1f4699f498","amount":"1000000000000000000","sendPortionEnabled":true,"type":"EXACT_INPUT","intent":"quote","configs":[{"useSyntheticQuotes":false,"routingType":"DUTCH_V2"},{"enableUniversalRouter":true,"protocols":["V2","V3","MIXED"],"routingType":"CLASSIC","enableFeeOnTransferFeeFetching":true}],"useUniswapX":true,"slippageTolerance":"0.5"}'


def sync_main():
    for i in range(10):
        s = requests.Session()
        s.headers.update(headers)
        response = s.post('https://interface.gateway.uniswap.org/v2/quote', data=data)
        print("requests ->", response.status_code)


async def async_main():
    for i in range(10):
        async with aiohttp.ClientSession(headers=headers) as session:
            async with session.request("POST",
                                       'https://interface.gateway.uniswap.org/v2/quote',
                                       data=data) as response:
                print("aiohttp ->", response.status)


if __name__ == '__main__':
    asyncio.run(async_main())
    sync_main()
```

Python 3.12
aiohttp==3.10.6
requests==2.32.3

**Comment by webknjaz** [comment, marked as answer]
You're being rate-limited: [`429 Too Many Requests`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429). Wait between the HTTP requests when you see a [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After).

**Reply by Dreamsorcerer** [reply]
Odd that it'd work with requests though. Do they have some automatic retry or something?
I wouldn't expect a noticeable difference in performance with that (unoptimised) code.

**Reply by webknjaz** [reply]
I'd assume that `requests` might be having a longer delay between requests, perhaps. Or maybe, the server sees different headers/TLS mode etc., and applies rate limiting differently.

**Reply by AndrejedaI** [reply]
I added this header, but unfortunately everything is the same, requests still returns 200, and aiohttp returns 429 after 6-7 requests. How can this happen in theory? I'm far from a professional in this, but this is very strange based on the fact that the resource and headers are the same

**Reply by AndrejedaI** [reply]
`'Retry-After': '5'` or is it a matter of delay?

**Reply by Dreamsorcerer** [reply]
You need to read the headers on the response, not add any...
